packages/eslint-plugin/src/rules/prefer-function-type.ts

```diff
@@ -26,6 +33,35 @@ export default util.createRule({
   create(context) {
     const sourceCode = context.getSourceCode();
 
+    /**
+     * Get the range of the `this` which is being used as a type annotation else returns null
+     * @param node The node being checked
+     * @returns {Array<number> | null} the range or null if no `this` type annotation are found
+     */
+    function getReturnType(
+      node: TSESTree.TSTypeAnnotation,
+    ): Array<number> | null {
+      if (!possibleReturnThisTypeHolders.has(node.type)) {
+        return null;
+      }
+
+      if (node.type === AST_NODE_TYPES.TSThisType) {
+        return node.range;
+      }
+
+      if (node.type === AST_NODE_TYPES.TSTypeAnnotation) {
+        if (node.typeAnnotation.type === AST_NODE_TYPES.TSThisType) {
+          return node.typeAnnotation.range;
+        }
+      }
+
+      if (node.type === AST_NODE_TYPES.TSFunctionType) {
+        return getReturnType(node?.returnType);
+      }
+
+      return null;
+    }
```

Comment:
```suggestion
    function getReturnType(
      node: TSESTree.TSTypeAnnotation,
    ): Array<number> | null {
      if (node.typeAnnotation.type === AST_NODE_TYPES.TSThisType) {
        return node.range;
      }

      if (node.typeAnnotation.type === AST_NODE_TYPES.TSFunctionType) {
        return getReturnType(node?.returnType);
      }
      
      // etc

      return null;
    }
```

Reply:
I tried this

```ts
    function getReturnType(
      node: TSESTree.TSTypeAnnotation,
    ): Array<number> | null {
      if (!possibleReturnThisTypeHolders.has(node.type)) {
        return null;
      }

      if (node.typeAnnotation.type === AST_NODE_TYPES.TSThisType) {
        return node.range;
      }

      if (node.typeAnnotation.type === AST_NODE_TYPES.TSTypeAnnotation) { // <--- type error
        if (node.typeAnnotation.type === AST_NODE_TYPES.TSThisType) {  // <--- type error
          return node.typeAnnotation.range;
        }
      }

      if (node.typeAnnotation.type === AST_NODE_TYPES.TSFunctionType) {
        return getReturnType(node?.returnType);
      }

      return null;
    }
```

Still no luck, even tests are failing cause of this !